agent/src/util/device_plugin_service.rs

```diff
@@ -894,6 +894,7 @@ mod device_plugin_service_tests {
             .collect();
         assert_eq!(devices.len(), capacity);
         // Can't use map on Device type
+        #[allow(clippy::needless_collect)]
```

Comment:
Might be optimal to refactor instead of adding flags where possible/not extensive.
I believe the following will work without the flag:
```
        let mut device_ids = devices.into_iter().map(|device| device.id);
        for device in expected_device_ids {
            assert!(device_ids.any(|d| d == device));
        }
```

Reply:
sorry @kate-goldenring thats I dont think gonna work. The `any` consumes the iterator (both into_iter and iter which i find odd, because `iter()` was supposed to be non consuming I thought). Here's an example if you wanna see what I mean: [gist](https://gist.github.com/Ragnyll/f77d24acb98a647b7d64c92c9b10e9d0) (this made me think i was going crazy for a sec).  So if the two sets of devices are not in the same order it'll eat up the non matching ones before it starts matching.

anyways reconstructing the iterator for each any in the loop works, and clippy doesnt complain: 8f48d04017fb60e55c4d90804072d8b533565666

if you dont like that i'd say just go back to the version with the `allow needless collect`,  this new way is kinda weird, so i'm not sure which is better.

Reply:
Really cool investigation -- this could be a blog post if not reason to put up a bug against the `allow needless collect` flag on clippy. Like you said, using the flag is neater more efficient code but what you have now is fine.